rust-mode.el

```diff
@@ -267,18 +268,40 @@ to the function arguments.  When nil, `->' will be indented one level."
         ;; Rewind until the point no longer moves
         (setq continue (/= starting (point)))))))
 
-(defun rust-in-macro ()
+(defvar-local rust-macro-scopes nil)
+
+(defun rust-macro-scope (&optional start end)
   (save-excursion
-    (when (> (rust-paren-level) 0)
-      (backward-up-list)
-      (rust-rewind-irrelevant)
-      (or (rust-looking-back-macro)
-          (and (rust-looking-back-ident)
-               (save-excursion
-                 (backward-sexp)
-                 (rust-rewind-irrelevant)
-                 (rust-looking-back-str "macro_rules!")))
-          (rust-in-macro)))))
+    ;; need to special case macro_rules which has unique syntax
+    (let ((scope nil))
+      (goto-char
+       (or start (point-min)))
+      ;; if there is a start move back to the first top level
+      (when start
+        (let ((top (syntax-ppss-toplevel-pos (syntax-ppss))))
+          (when top
+            (goto-char top))))
+      (while (search-forward "!" end t)
+        (when (and (not (rust-in-str-or-cmnt))
+                   (skip-chars-forward " \t\n\r")
+                   (or
+                    (eq ?\[ (char-after))
+                    (eq ?\( (char-after))
+                    (eq ?\{ (char-after))))
+          (let ((start (point)))
+            (forward-list)
+            (setq scope (cons (list start (point)) scope)))))
+      scope)))
+
+(defun rust-in-macro (&optional start end)
+  (when (> (rust-paren-level) 0)
+    (let ((scopes
+           (or rust-macro-scopes (rust-macro-scope start end))))
+      (seq-some
+       (lambda (sc)
+         (and (>= (point) (car sc))
```

Comment:
how does it compare to a binary search?

Reply:
Haven't timed it. Probably not that much difference because lists are linear time access. I really wouldn't be too worried about manipulation of integer lists, though, compared to all the buffer movement.

Reply:
Sure, binary search should be done on an array structure (or a binary tree kind of data structure). It could still be done for another PR, don't bother with that just yet.